File: core/src/main/java/org/iumotionlabs/hefesto/feature/portinfo/output/JsonFormatter.java
package org.iumotionlabs.hefesto.feature.portinfo.output;

import org.iumotionlabs.hefesto.feature.portinfo.model.*;
import tools.jackson.databind.ObjectMapper;
import tools.jackson.databind.SerializationFeature;
import tools.jackson.databind.json.JsonMapper;
import tools.jackson.databind.node.ArrayNode;
import tools.jackson.databind.node.ObjectNode;

import java.util.List;
import java.util.Map;

/**
 * Formats output as JSON with enhanced structure.
 */
public final class JsonFormatter implements OutputFormatter {

    private final ObjectMapper mapper;
    private final boolean pretty;

    public JsonFormatter() {
        this(true);
    }

    public JsonFormatter(boolean pretty) {
        this.pretty = pretty;
        JsonMapper.Builder builder = JsonMapper.builder();
        if (pretty) {
            builder.enable(SerializationFeature.INDENT_OUTPUT);
        }
        this.mapper = builder.build();
    }

    @Override
    public String format(List<EnrichedPortBinding> bindings) {
        try {
            ArrayNode array = mapper.createArrayNode();

            for (EnrichedPortBinding b : bindings) {
                ObjectNode node = mapper.createObjectNode();
                node.put("port", b.port());
                node.put("protocol", b.protocol());
                node.put("state", b.state());
                node.put("localAddress", b.localAddress());
                node.put("remoteAddress", b.binding().remoteAddress());
                node.put("remotePort", b.binding().remotePort());
                node.put("pid", b.pid());
                node.put("processName", b.processName());
                node.put("user", b.user());
                node.put("commandLine", b.binding().commandLine());
                node.put("isExposed", b.isExposed());
                node.put("isLocalOnly", b.isLocalOnly());

                if (b.serviceInfo() != null) {
                    ObjectNode serviceNode = mapper.createObjectNode();
                    serviceNode.put("name", b.serviceInfo().name());
                    serviceNode.put("description", b.serviceInfo().description());
                    serviceNode.put("category", b.serviceInfo().category().name());
                    node.set("service", serviceNode);
                }

                if (b.dockerInfo() != null) {
                    ObjectNode dockerNode = mapper.createObjectNode();
                    dockerNode.put("containerId", b.dockerInfo().shortId());
                    dockerNode.put("containerName", b.dockerInfo().containerName());
                    dockerNode.put("image", b.dockerInfo().image());
                    dockerNode.put("status", b.dockerInfo().status());
                    node.set("docker", dockerNode);
                }

                array.add(node);
            }

            return mapper.writeValueAsString(array);
        } catch (Exception e) {
            return "[]";
        }
    }

    @Override
    public String formatOverview(NetworkOverview overview) {
        try {
            ObjectNode root = mapper.createObjectNode();

            // Statistics
            ObjectNode stats = mapper.createObjectNode();
            stats.put("totalListening", overview.statistics().totalListening());
            stats.put("totalEstablished", overview.statistics().totalEstablished());
            stats.put("tcpCount", overview.statistics().tcpCount());
            stats.put("udpCount", overview.statistics().udpCount());
            stats.put("exposedCount", overview.statistics().exposedCount());
            stats.put("localOnlyCount", overview.statistics().localOnlyCount());
            root.set("statistics", stats);

            // Timestamp
            root.put("generatedAt", overview.generatedAt().toString());

            // Bindings grouped by process
            ObjectNode processes = mapper.createObjectNode();
            for (Map.Entry<Long, List<EnrichedPortBinding>> entry : overview.byPid().entrySet()) {
                List<EnrichedPortBinding> ports = entry.getValue();
                if (ports.isEmpty()) continue;

                EnrichedPortBinding first = ports.get(0);
                ObjectNode processNode = mapper.createObjectNode();
                processNode.put("pid", first.pid());
                processNode.put("name", first.processName());
                processNode.put("user", first.user());

                ArrayNode portsArray = mapper.createArrayNode();
                for (EnrichedPortBinding b : ports) {
                    ObjectNode portNode = mapper.createObjectNode();
                    portNode.put("port", b.port());
                    portNode.put("protocol", b.protocol());
                    portNode.put("state", b.state());
                    portNode.put("address", b.localAddress());
                    portNode.put("exposed", b.isExposed());
                    if (b.serviceInfo() != null) {
                        portNode.put("service", b.serviceInfo().name());
                    }
                    portsArray.add(portNode);
                }
                processNode.set("ports", portsArray);

                processes.set(String.valueOf(entry.getKey()), processNode);
            }
            root.set("processes", processes);

            // Exposed ports summary
            ArrayNode exposed = mapper.createArrayNode();
            for (EnrichedPortBinding b : overview.exposedPorts()) {
                ObjectNode node = mapper.createObjectNode();
                node.put("port", b.port());
                node.put("process", b.processName());
                node.put("pid", b.pid());
                if (b.serviceInfo() != null) {
                    node.put("service", b.serviceInfo().name());
                }
                exposed.add(node);
            }
            root.set("exposedPorts", exposed);

            return mapper.writeValueAsString(root);
        } catch (Exception e) {
            return "{}";
        }
    }

    @Override
    public String formatSecurityReport(SecurityReport report) {
        try {
            ObjectNode root = mapper.createObjectNode();

            // Summary
            ObjectNode summary = mapper.createObjectNode();
            summary.put("critical", report.criticalCount());
            summary.put("high", report.highCount());
            summary.put("warning", report.warningCount());
            summary.put("info", report.infoCount());
            summary.put("totalPortsAnalyzed", report.totalPortsAnalyzed());
            summary.put("clean", report.isClean());
            root.set("summary", summary);

            root.put("generatedAt", report.generatedAt().toString());

            // Findings
            ArrayNode findings = mapper.createArrayNode();
            for (SecurityFlag flag : report.sortedBySeverity()) {
                ObjectNode node = mapper.createObjectNode();
                node.put("severity", flag.severity().name());
                node.put("category", flag.category().name());
                node.put("title", flag.title());
                node.put("description", flag.description());
                node.put("recommendation", flag.recommendation());

                if (flag.relatedBinding() != null) {
                    ObjectNode binding = mapper.createObjectNode();
                    binding.put("port", flag.relatedBinding().port());
                    binding.put("protocol", flag.relatedBinding().protocol());
                    binding.put("process", flag.relatedBinding().processName());
                    binding.put("pid", flag.relatedBinding().pid());
                    binding.put("address", flag.relatedBinding().localAddress());
                    node.set("binding", binding);
                }

                findings.add(node);
            }
            root.set("findings", findings);

            return mapper.writeValueAsString(root);
        } catch (Exception e) {
            return "{}";
        }
    }

    @Override
    public String formatHealthCheck(HealthCheckResult result) {
        try {
            ObjectNode root = mapper.createObjectNode();

            root.put("port", result.port());
            root.put("protocol", result.protocol());
            root.put("status", result.status().name());
            root.put("statusDisplay", result.status().displayName());
            root.put("healthy", result.isHealthy());
            root.put("responseTimeMs", result.responseTimeMs());
            root.put("message", result.message());
            root.put("timestamp", result.timestamp().toString());

            result.httpInfoOpt().ifPresent(http -> {
                ObjectNode httpNode = mapper.createObjectNode();
                httpNode.put("statusCode", http.statusCode());
                httpNode.put("statusText", http.statusText());
                httpNode.put("contentType", http.contentType());
                httpNode.put("contentLength", http.contentLength());
                httpNode.put("responseTimeMs", http.responseTimeMs());
                httpNode.put("success", http.isSuccess());
                root.set("http", httpNode);
            });

            result.sslInfoOpt().ifPresent(ssl -> {
                ObjectNode sslNode = mapper.createObjectNode();
                sslNode.put("issuer", ssl.issuer());
                sslNode.put("subject", ssl.subject());
                sslNode.put("validFrom", ssl.validFrom() != null ? ssl.validFrom().toString() : null);
                sslNode.put("validTo", ssl.validTo() != null ? ssl.validTo().toString() : null);
                sslNode.put("protocol", ssl.protocol());
                sslNode.put("cipherSuite", ssl.cipherSuite());
                sslNode.put("valid", ssl.valid());
                sslNode.put("expired", ssl.isExpired());
                sslNode.put("expiresSoon", ssl.expiresSoon());
                sslNode.put("daysUntilExpiry", ssl.daysUntilExpiry());
                root.set("ssl", sslNode);
            });

            return mapper.writeValueAsString(root);
        } catch (Exception e) {
            return "{}";
        }
    }

    @Override
    public String mimeType() {
        return "application/json";
    }

    @Override
    public String fileExtension() {
        return "json";
    }
}

File: hefesto-core/src/feature/portinfo/output/json_formatter.rs
use serde_json::{json, Map, Value};

use hefesto_domain::portinfo::enriched_port_binding::EnrichedPortBinding;
use hefesto_domain::portinfo::health_check::HealthCheckResult;
use hefesto_domain::portinfo::network_overview::NetworkOverview;
use hefesto_domain::portinfo::security::SecurityReport;

use super::OutputFormatter;

/// Formats output as JSON with enhanced structure.
///
/// All numeric values are formatted in a locale-independent manner (Rust
/// guarantees this by default -- no locale-dependent number formatting).
pub struct JsonFormatter {
    pretty: bool,
}

impl JsonFormatter {
    /// Creates a new `JsonFormatter` with pretty printing.
    pub fn new() -> Self {
        Self { pretty: true }
    }

    /// Creates a new `JsonFormatter` with optional pretty printing.
    pub fn with_pretty(pretty: bool) -> Self {
        Self { pretty }
    }

    fn to_string(&self, value: &Value) -> String {
        if self.pretty {
            serde_json::to_string_pretty(value).unwrap_or_else(|_| "{}".to_string())
        } else {
            serde_json::to_string(value).unwrap_or_else(|_| "{}".to_string())
        }
    }
}

impl Default for JsonFormatter {
    fn default() -> Self {
        Self::new()
    }
}

impl OutputFormatter for JsonFormatter {
    fn format(&self, bindings: &[EnrichedPortBinding]) -> String {
        let array: Vec<Value> = bindings
            .iter()
            .map(|b| {
                let mut obj = json!({
                    "port": b.port(),
                    "protocol": b.protocol().as_str(),
                    "state": b.state().as_str(),
                    "localAddress": b.local_address(),
                    "remoteAddress": b.binding.remote_address,
                    "remotePort": b.binding.remote_port,
                    "pid": b.pid(),
                    "processName": b.process_name(),
                    "user": b.user(),
                    "commandLine": b.binding.command_line,
                    "isExposed": b.is_exposed(),
                    "isLocalOnly": b.is_local_only(),
                });

                if let Some(ref si) = b.service_info {
                    obj["service"] = json!({
                        "name": si.name,
                        "description": si.description,
                        "category": format!("{:?}", si.category),
                    });
                }

                if let Some(ref di) = b.docker_info {
                    obj["docker"] = json!({
                        "containerId": di.short_id(),
                        "containerName": di.container_name,
                        "image": di.image,
                        "status": di.status,
                    });
                }

                obj
            })
            .collect();

        let value = Value::Array(array);
        self.to_string(&value)
    }

    fn format_overview(&self, overview: &NetworkOverview) -> String {
        let stats = &overview.statistics;
        let mut root = json!({
            "statistics": {
                "totalListening": stats.total_listening,
                "totalEstablished": stats.total_established,
                "tcpCount": stats.tcp_count,
                "udpCount": stats.udp_count,
                "exposedCount": stats.exposed_count,
                "localOnlyCount": stats.local_only_count,
            },
            "generatedAt": overview.generated_at.to_rfc3339(),
        });

        // Bindings grouped by process
        let by_pid = overview.by_pid();
        let mut processes = Map::new();

        for (&pid, ports) in &by_pid {
            if ports.is_empty() {
                continue;
            }
            let first = ports[0];

            let ports_array: Vec<Value> = ports
                .iter()
                .map(|b| {
                    let mut port_obj = json!({
                        "port": b.port(),
                        "protocol": b.protocol().as_str(),
                        "state": b.state().as_str(),
                        "address": b.local_address(),
                        "exposed": b.is_exposed(),
                    });
                    if let Some(ref si) = b.service_info {
                        port_obj["service"] = json!(si.name);
                    }
                    port_obj
                })
                .collect();

            let process_node = json!({
                "pid": first.pid(),
                "name": first.process_name(),
                "user": first.user(),
                "ports": ports_array,
            });

            processes.insert(pid.to_string(), process_node);
        }

        root["processes"] = Value::Object(processes);

        // Exposed ports summary
        let exposed: Vec<Value> = overview
            .exposed_ports()
            .iter()
            .map(|b| {
                let mut node = json!({
                    "port": b.port(),
                    "process": b.process_name(),
                    "pid": b.pid(),
                });
                if let Some(ref si) = b.service_info {
                    node["service"] = json!(si.name);
                }
                node
            })
            .collect();

        root["exposedPorts"] = Value::Array(exposed);

        self.to_string(&root)
    }

    fn format_security_report(&self, report: &SecurityReport) -> String {
        let root = json!({
            "summary": {
                "critical": report.critical_count(),
                "high": report.high_count(),
                "warning": report.warning_count(),
                "info": report.info_count(),
                "totalPortsAnalyzed": report.total_ports_analyzed,
                "clean": report.is_clean(),
            },
            "generatedAt": report.generated_at.to_rfc3339(),
            "findings": report.sorted_by_severity().iter().map(|flag| {
                json!({
                    "severity": format!("{:?}", flag.severity),
                    "category": format!("{:?}", flag.category),
                    "title": flag.title,
                    "description": flag.description,
                    "recommendation": flag.recommendation,
                    "binding": {
                        "port": flag.related_port,
                        "process": flag.related_process,
                    },
                })
            }).collect::<Vec<Value>>(),
        });

        self.to_string(&root)
    }

    fn format_health_check(&self, result: &HealthCheckResult) -> String {
        let mut root = json!({
            "port": result.port,
            "protocol": result.protocol,
            "status": format!("{:?}", result.status),
            "statusDisplay": result.status.display_name(),
            "healthy": result.is_healthy(),
            "responseTimeMs": result.response_time_ms,
            "message": result.message,
            "timestamp": result.timestamp.to_rfc3339(),
        });

        if let Some(ref http) = result.http_info {
            root["http"] = json!({
                "statusCode": http.status_code,
                "statusText": http.status_text,
                "contentType": http.content_type,
                "contentLength": http.content_length,
                "responseTimeMs": http.response_time_ms,
                "success": http.is_success(),
            });
        }

        if let Some(ref ssl) = result.ssl_info {
            root["ssl"] = json!({
                "issuer": ssl.issuer,
                "subject": ssl.subject,
                "validFrom": ssl.valid_from.map(|dt| dt.to_rfc3339()),
                "validTo": ssl.valid_to.map(|dt| dt.to_rfc3339()),
                "protocol": ssl.protocol,
                "cipherSuite": ssl.cipher_suite,
                "valid": ssl.valid,
                "expired": ssl.is_expired(),
                "expiresSoon": ssl.expires_soon(),
                "daysUntilExpiry": ssl.days_until_expiry(),
            });
        }

        self.to_string(&root)
    }

    fn mime_type(&self) -> &str {
        "application/json"
    }

    fn file_extension(&self) -> &str {
        "json"
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use hefesto_domain::portinfo::health_check::HealthStatus;
    use hefesto_domain::portinfo::port_binding::{PortBinding, Protocol};
    use hefesto_domain::portinfo::security::SecurityReport;
    use hefesto_domain::portinfo::service_info::{ServiceCategory, ServiceInfo};

    #[test]
    fn test_format_empty() {
        let formatter = JsonFormatter::new();
        let result = formatter.format(&[]);
        assert_eq!(result.trim(), "[]");
    }

    #[test]
    fn test_format_single_binding() {
        let formatter = JsonFormatter::with_pretty(false);
        let binding = PortBinding::listen(8080, Protocol::Tcp, 1234, "java");
        let enriched = EnrichedPortBinding::from_binding(binding);
        let result = formatter.format(&[enriched]);

        let parsed: Value = serde_json::from_str(&result).unwrap();
        assert!(parsed.is_array());
        assert_eq!(parsed[0]["port"], 8080);
        assert_eq!(parsed[0]["processName"], "java");
    }

    #[test]
    fn test_format_with_service_info() {
        let formatter = JsonFormatter::with_pretty(false);
        let binding = PortBinding::listen(3306, Protocol::Tcp, 100, "mysqld");
        let enriched = EnrichedPortBinding::with_service(
            binding,
            ServiceInfo::new("MySQL", "MySQL Database", ServiceCategory::Database),
        );
        let result = formatter.format(&[enriched]);

        let parsed: Value = serde_json::from_str(&result).unwrap();
        assert_eq!(parsed[0]["service"]["name"], "MySQL");
    }

    #[test]
    fn test_format_health_check() {
        let formatter = JsonFormatter::with_pretty(false);
        let result = HealthCheckResult::tcp(8080, HealthStatus::Reachable, 42, "OK");
        let output = formatter.format_health_check(&result);

        let parsed: Value = serde_json::from_str(&output).unwrap();
        assert_eq!(parsed["port"], 8080);
        assert_eq!(parsed["healthy"], true);
        assert_eq!(parsed["responseTimeMs"], 42);
    }

    #[test]
    fn test_format_security_clean() {
        let formatter = JsonFormatter::with_pretty(false);
        let report = SecurityReport::empty();
        let output = formatter.format_security_report(&report);

        let parsed: Value = serde_json::from_str(&output).unwrap();
        assert_eq!(parsed["summary"]["clean"], true);
    }

    #[test]
    fn test_locale_independent_numbers() {
        // Rust does not use locale for number formatting, but verify explicitly
        let formatter = JsonFormatter::with_pretty(false);
        let result = HealthCheckResult::tcp(8080, HealthStatus::Reachable, 12345, "OK");
        let output = formatter.format_health_check(&result);
        assert!(output.contains("12345"));
        // Ensure no commas as thousand separators
        assert!(!output.contains("12,345"));
    }
}
